src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/metrics/MetricsCollectorSpark.java

```diff
@@ -14,41 +14,72 @@
+ *   XMetrics extends MetricsBase: defines the aggregate metrics that we're trying to collect
+ *   XMetricsArgumentCollection: defines parameters for XMetrics, extends MetricsArgumentCollection
+ *   XMetricsCollector: processes a single read, and has a reduce/combiner
+ *       For multi level collectors, XMetricsCollector is composed of several classes:
+ *           XMetricsCollector extends MultiLevelCollectorReducible<
+ *              XMetrics, HISTOGRAM_KEY, XMetricsCollectorArgs, XMetricsPerUnitCollector>
+ *           XMetricsPerUnitCollector: per level collector, implements
+ *              PerUnitCollector<XMetrics, HISTOGRAM_KEY, XMetricsCollectorArgs> (requires a combiner)
+ *           XMetricsCollectorArgs per-record argument (type argument for MultiLevelCollector)
+ *   XMetricsCollectorSpark: adapter/bridge between RDD and the (read-based) XMetricsCollector,
+ *          implements MetricsCollectorSpark<XMetricsArgumentCollection>
+ *   CollectXMetrics extends SinglePassProgram
+ *   CollectXMetricsSpark extends MetricsCollectorSparkTool<MyMetricsArgumentCollection>
  *
- * The general lifecycle of a Spark collector looks like this:
+ * The following schematic shows the general relationships of these collector component classes
+ * in the context of various tools, with the arrows indicating a "delegates to" relationship
+ * via composition or inheritance:
+ *
+ *     CollectXMetrics    CollectMultipleMetrics
+ *                \               /
+ *                \              /
+ *                v             v
+ *      _______________________________________
+ *      |          XMetricsCollector =========|=========> MultiLevelCollectorReducible
+ *      |                 |                   |                      |
+ *      |                 V                   |                      |
+ *      |             XMetrics                |                      V
+ *      | XMetricsCollectorArgumentCollection |             PerUnitXMetricCollector
+ *      ---------------------------------------
+ *                       ^
+ *                       |
+ *                       |
+ *             XMetricsCollectorSpark
+ *                ^              ^
+ *               /               \
+ *              /                \
+ *    CollectXMetricsSpark  CollectMultipleMetricsSpark
+ *
+ *
+ * The general lifecycle of a Spark collector (XMetricsCollectorSpark in the diagram
+ * above) looks like this:
  *
  *     CollectorType collector = new CollectorType<CollectorArgType>()
  *     CollectorArgType args = // get metric-specific input arguments
  *
- *     // pass the input arguments back to the collector for initialization
+ *     // pass the input arguments to the collector for initialization
  *     collector.initialize(args);
  *
- *     ReadFilter filter == collector.getReadFilter();
+ *     ReadFilter filter == collector.getReadFilter(samFileHeader);
```

Comment:
is this `samFileHeader` different from `getHeaderForReads()` 2 lines below? I think it's the same